File: tools/lib/vault-config.js
/**
 * Vault Config - Reads credentials from DPAPI vault
 *
 * Usage:
 *   const { getSmtpConfig, getImapConfig, getCredential } = require('./lib/vault-config');
 *   const smtp = getSmtpConfig();
 *   const imap = getImapConfig();
 */

const { execSync } = require('child_process');
const path = require('path');

const VAULT_SCRIPT = path.join(__dirname, '..', 'vault.ps1');

// Cache to avoid repeated vault calls in same process
const _cache = {};

function vaultGet(service, field) {
  const cacheKey = `${service}:${field}`;
  if (_cache[cacheKey]) return _cache[cacheKey];

  try {
    const cmd = `powershell.exe -NoProfile -ExecutionPolicy Bypass -File "${VAULT_SCRIPT}" -Action get -Service "${service}" -Field "${field}" -Silent`;
    const result = execSync(cmd, { encoding: 'utf8', timeout: 10000 }).trim();
    if (!result || result === '[DECRYPTION FAILED]') {
      throw new Error(`Vault returned empty or failed for ${service}:${field}`);
    }
    _cache[cacheKey] = result;
    return result;
  } catch (err) {
    console.error(`[vault-config] Failed to get ${service}:${field} - ${err.message}`);
    process.exit(1);
  }
}

function getSmtpConfig() {
  return {
    host: 'mail.zxcs.nl',
    port: 465,
    secure: true,
    auth: {
      user: vaultGet('smtp', 'username'),
      pass: vaultGet('smtp', 'password')
    },
    tls: { rejectUnauthorized: false },
    connectionTimeout: 10000,
    greetingTimeout: 10000
  };
}

function getImapConfig() {
  return {
    user: vaultGet('imap', 'username'),
    password: vaultGet('imap', 'password'),
    host: 'mail.zxcs.nl',
    port: 993,
    tls: true,
    tlsOptions: { rejectUnauthorized: false },
    authTimeout: 30000,
    connTimeout: 30000
  };
}

function getCredential(service, field) {
  return vaultGet(service, field);
}

module.exports = { getSmtpConfig, getImapConfig, getCredential, vaultGet };

File: tools/lib/vault-config.py
"""
Vault Config - Reads credentials from DPAPI vault for Python scripts.

Usage:
    from lib.vault_config import get_imap_credentials, get_credential
    user, password = get_imap_credentials()
"""

import subprocess
import os

VAULT_SCRIPT = os.path.join(os.path.dirname(__file__), '..', 'vault.ps1')

_cache = {}

def vault_get(service, field):
    cache_key = f"{service}:{field}"
    if cache_key in _cache:
        return _cache[cache_key]

    cmd = [
        'powershell.exe', '-NoProfile', '-ExecutionPolicy', 'Bypass',
        '-File', VAULT_SCRIPT,
        '-Action', 'get', '-Service', service, '-Field', field, '-Silent'
    ]
    result = subprocess.run(cmd, capture_output=True, timeout=10)
    value = result.stdout.decode('utf-8').strip()

    if not value or value == '[DECRYPTION FAILED]':
        raise RuntimeError(f"Vault returned empty or failed for {service}:{field}")

    _cache[cache_key] = value
    return value


def get_imap_credentials():
    """Returns (username, password) for IMAP."""
    return vault_get('imap', 'username'), vault_get('imap', 'password')


def get_smtp_credentials():
    """Returns (username, password) for SMTP."""
    return vault_get('smtp', 'username'), vault_get('smtp', 'password')


def get_credential(service, field):
    """Generic credential retrieval."""
    return vault_get(service, field)
